### ISSUE TITLE
[Feature Request] Highlight invalid links in red

### ISSUE THREAD

**Issue by S-ven-Peters** [gollum/gollum#1965]
Currently when viewing or editing an article invalid links look identical to valid links.
Therefore, all links look valid.
If invalid links would be shown in red, it would allow for detection of invalid links.

So my suggestion is:
- style invalid links in red when reading an article
- style invalid links in red when previewing an article from edit
- highlight invalid links' markup text in red when editing an article

**Comment by benjaminwil** [comment]
Hey, this is a cool idea.

There isn't much we'd be able to reasonably to about "invalid" external links to valid URLs (i.e. `https://example.com/this-page-does-not-exist-but-is-a-valid-url.html`) but we could hypothetically implement this for internal wiki links:

- For links that are certainly invalid for some reason.
- For links to pages that do not exist yet.

It sounds like you are inspired by how Wikipedia works, where red links are for pages that do not exist. I'm not sure how Wikipedia handles "invalid" links, as I would not call a link to a page that doesn't exist invalid, personally.

If someone were interested in picking up this feature work, I'd be happy to provide feedback on pull requests or discuss more on this issue. 👍

**Comment by guillaume-d** [comment]
> Currently when viewing or editing an article invalid links look identical to valid links. Therefore, all links look valid. If invalid links would be shown in red, it would allow for detection of invalid links.
> 
> So my suggestion is:
> 
>     * style invalid links in red when reading an article
> 

As far as I can understand from the code, this is done for the Gollum-specific double-bracketed link syntax (for example `[[NonExistingPage]]`), see https://github.com/gollum/gollum-lib/issues/436#issuecomment-1435948133.
I would also be very happy if I do not have to implement this myself in Javascript for Markdown, as explained there.

>     * style invalid links in red when previewing an article from edit
> 
>     * highlight invalid links' markup text in red when editing an article

Great ideas a well! :+1:

**Comment by S-ven-Peters** [comment]
It does indeed work with the gollum specific syntax. In the article view and in the preview as well.

**Comment by dometto** [comment]
> It does indeed work with the gollum specific syntax. In the article view and in the preview as well.

Indeed! Gollum has no control over the way different renderers render links, so we can't make them check whether pages exist. I think any other solution would require checking each link and seeing whether it is a valid page (or a redirect to `/gollum/create`), which doesn't seem like a nice approach at all. So unless anyway has a good idea for how to implement this cleanly and in a way that's easy to maintain, I unfortunately have to suggest closing this issue.

**Comment by guillaume-d** [comment]
The best I could think of so far was the following:

Define an HTTP API endpoint that would take a list of unique document paths and return a bit array where the values depend on whether the corresponding file exists or not.

Use that from Javascript to recolor the links asynchronously, batching the paths in as many chunks as needed to find the right balance between speed and UI responsiveness.

Lots of flickering and inefficiency to expect, but still better than nothing, and relatively uninvasive.
